src/web3/KeychainLock.ts

```diff
@@ -0,0 +1,254 @@
+
+    // Check if we need to migrate the existing Geth KeyStore based account into the keychain, by importing it from the mnemonic
+    if (importMnemonicAccount?.address) {
+      const normalizedMnemonicAccountAddress = normalizeAddressWith0x(importMnemonicAccount.address)
+      const account = accounts.find((a) => a.address === normalizedMnemonicAccountAddress)
+      if (!account) {
+        Logger.info(
+          `${TAG}@listStoredAccounts`,
+          `Existing account ${importMnemonicAccount.address} not found in the keychain, will import from the stored mnemonic on next unlock`
+        )
+        const now = Date.now()
+        // Ensure the imported account is always first in the list
+        const maxCreationTime = Math.min(
+          importMnemonicAccount.createdAt.getTime(),
+          (accounts[0]?.createdAt.getTime() ?? now) - 1,
+          now
+        )
+        const createdAt = new Date(maxCreationTime)
+        accounts = [
+          { address: normalizedMnemonicAccountAddress, createdAt, importFromMnemonic: true },
+          ...accounts,
+        ]
+      }
+    }
+  } catch (e) {
+    Logger.error(`${TAG}@listStoredAccounts`, 'Error listing accounts', e)
+    throw new Error(ErrorMessages.KEYCHAIN_FETCH_ACCOUNTS)
+  }
+  return accounts
+}
+
+async function importAndStorePrivateKeyFromMnemonic(account: KeychainAccount, password: string) {
+  const mnemonic = await getStoredMnemonic(account.address, password)
+  if (!mnemonic) {
+    throw new Error('No mnemonic found in storage')
+  }
+
+  const { privateKey } = await generateKeysFromMnemonic(mnemonic)
+  if (!privateKey) {
+    throw new Error('Failed to generate private key from mnemonic')
+  }
+
+  // Prefix 0x here or else the signed transaction produces dramatically different signer!!!
+  const normalizedPrivateKey = normalizeAddressWith0x(privateKey)
+  const accountFromPrivateKey = normalizeAddressWith0x(privateKeyToAddress(normalizedPrivateKey))
+  if (accountFromPrivateKey !== account.address) {
+    throw new Error(
+      `Generated private key address (${accountFromPrivateKey}) does not match the existing account address (${account.address})`
+    )
+  }
+
+  await storePrivateKey(normalizedPrivateKey, account, password)
+
+  return normalizedPrivateKey
+}
+
+export async function clearStoredAccounts() {
+  Logger.info(`${TAG}@clearStoredAccounts`, 'Clearing all accounts from keychain')
+  const accounts = await listStoredAccounts()
+  await Promise.all(accounts.map((account) => removeStoredItem(accountStorageKey(account))))
+}
+
+export class KeychainLock {
+  protected locks: Map<
+    string,
+    {
+      // Timestamp in milliseconds when the keychain was last unlocked
+      unlockTime?: number
+      // Number of seconds that the keychain was last unlocked for
+      unlockDuration?: number
+      account: KeychainAccount
+    }
+  > = new Map()
+
+  addAccount(account: KeychainAccount) {
```

Comment:
If `addAccount` is called multiple times for the same account address and the account was previously unlocked, this will (in effect) lock the account prematurely -- is this a possibility?

Reply:
It should be impossible from the KeychainSigner/Wallet side because they have checks before calling addAccount to see if the account is already in the keychain. I'll add a the same check here though to make sure that functionality is preserved.

Reply:
Strike that, it seems it is already possible for this to be called twice, similarly it is possible for [addSigner](https://github.com/celo-org/celo-monorepo/blob/0dfd5730c19cb0b96cf05c7cff3e82d2f1f109cd/packages/sdk/wallets/wallet-base/src/wallet-base.ts#L59) to be called twice so I'm keeping the implementation as-is similar to that.